xbmc/platform/win10/AsyncHelpers.h

```diff
@@ -24,69 +24,70 @@
 #include <ppl.h>
 #include <ppltasks.h>
 
-inline void Wait(Windows::Foundation::IAsyncAction^ asyncOp)
+namespace winrt
 {
+  using namespace Windows::Foundation;
+}
+
+inline void Wait(winrt::IAsyncAction const& asyncOp)
+{
+  if (asyncOp.Status() == winrt::AsyncStatus::Completed)
+    return;
+
   auto __sync = std::make_shared<Concurrency::event>();
-  asyncOp->Completed = ref new Windows::Foundation::AsyncActionCompletedHandler(
-    [&](Windows::Foundation::IAsyncAction^ op, Windows::Foundation::AsyncStatus st)
-  {
+  asyncOp.Completed([&](auto&&, auto&&) {
```

Comment:
I'm curious if we could use co_async and co_await in the windows specific bits to simplify the async operations.
Think async/await but for C++. It's implemented and supported for production work in msvc and implemented in clang, both implementations by the same guy. Should hopefully make it into C++20 if we're lucky.

Reply:
In most cases we are using asynchronous WinRT APIs in synchronous context. that's why I've implemented this helpers. Unfortunately, `IAsyncOperation::get()` method suspend UI thread what isn't allowed in WinRT.